api/v1beta1/openstacknet_types.go

```diff
@@ -0,0 +1,158 @@
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OpenStackNetSpec defines the desired state of OpenStackNet
+type OpenStackNetSpec struct {
+	// +kubebuilder:validation:Optional
+	//  List of subnets
+	Subnets []OpenStackSubnet `json:"subnets,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1500
+	// MTU of the network
+	MTU int `json:"mtu"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// VIP create virtual ip on the network
+	VIP bool `json:"vip"`
+
+	// +kubebuilder:validation:Required
+	// DomainName the name of the domain for this network, usually lower(Name)."OSNetConfig.Spec.DomainName"
+	DomainName string `json:"domainName"`
+}
+
+// OpenStackSubnet type
+type OpenStackSubnet struct {
+
+	// +kubebuilder:validation:Required
+	// Name of the subnet
+	Name string `json:"name"`
+
+	// +kubebuilder:validation:Required
+	// Cidr the cidr to use for this network
+	Cidr string `json:"cidr"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum=4094
+	// Vlan ID
+	Vlan int `json:"vlan,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// AllocationRange a set of IPs for assignment
+	AllocationRange AllocationRange `json:"allocationRange"`
+
+	// +kubebuilder:validation:Optional
+	// Gateway optional gateway for the network
+	Gateway string `json:"gateway,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Routes, list of networks that should be routed via network gateway.
+	Routes []Route `json:"routes,omitempty"`
+}
+
+// AllocationRange type
+type AllocationRange struct {
+	// +kubebuilder:validation:Required
+	// AllocationStart a set of IPs that are reserved and will not be assigned
+	AllocationStart string `json:"allocationStart"`
+
+	// +kubebuilder:validation:Required
+	// AllocationEnd a set of IPs that are reserved and will not be assigned
+	AllocationEnd string `json:"allocationEnd"`
+}
+
+// Route definition
+type Route struct {
+	// +kubebuilder:validation:Required
+	// Destination, network CIDR
+	Destination string `json:"destination"`
+
+	// +kubebuilder:validation:Required
+	// Nexthop, gateway for the destination
+	Nexthop string `json:"nexthop"`
+}
+
+// IPReservation contains an IP and Deleted flag
+type IPReservation struct {
+	// +kubebuilder:validation:Required
+	IP string `json:"ip"`
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	VIP bool `json:"vip"`
+	// +kubebuilder:validation:Optional
+	Gateway string `json:"gateway,omitempty"`
+	// +kubebuilder:validation:Optional
+	Routes []Route `json:"routes,omitempty"`
+}
+
+// OpenStackNetStatus defines the observed state of OpenStackNet
+type OpenStackNetStatus struct {
+	// Already reserved IPs
+	// +kubebuilder:validation:Optional
+	Reservations []IPReservation `json:"reservations,omitempty"`
```

Comment:
how would we handle backup/restore with the Reservations stored in status? you can not just save the yaml of an osnet with its status and apply it back to the env. it would required special restore to also set the status to a state. thats why when we used the osnet in OSPDO we stored the reservations in the spec and a reservation was successfully stored if it was written to the spec, but also when the osnet controller successfully has updated its status with a different reservation format.